include/dca/parallel/mpi_concurrency/mpi_collective_max.hpp

```diff
@@ -23,20 +24,20 @@ namespace parallel {
 
 class MPICollectiveMax {
 public:
-  MPICollectiveMax(const MPIProcessorGrouping& grouping) : grouping_(grouping) {}
+  MPICollectiveMax(const std::unique_ptr<const MPIProcessorGrouping>& grouping) : grouping_(grouping) {}
 
   template <typename Scalar>
   void max(Scalar& value) const {
     Scalar result;
 
     MPI_Allreduce(&value, &result, MPITypeMap<Scalar>::factor(), MPITypeMap<Scalar>::value(),
-                  MPI_MAX, grouping_.get());
+                  MPI_MAX, grouping_->get());
 
     value = result;
   }
 
 private:
-  const MPIProcessorGrouping& grouping_;
+  const std::unique_ptr<const MPIProcessorGrouping>& grouping_;
 };
 
 }  // parallel
```

Comment:
The designed here is creative and I'm not sure necessary, but @gbalduzz may be able to argue for its usefulness better since it is his. design.

As I understand it:
These pointer references come from the derived MPIConcurrency class which gains its difference collective methods by multiple inheritance of the difference mpi collective base implementations. The desire here is to have access to this unique pointer of the derived class in the base class. Which may be reset during the lifetime of the derived Concurrency object. The derived class controls the lifetime of the object it has the actual pointer. We don't want to give the base class std::shared's because that will not update with the std::unique_ptr of the derived class.

I don't think it would be possible to use the stale reference since access to the base class methods occurs through the derived class to which the processor grouping lifetime is tied. If you acquired and accessed the collective method through a pointer to the base mpi collective I imagine horrible things could occur.

Reply:
The motivation is that  MPI_concurrency needs to initiate MPI before creating a grouping object, then all of the concurrency's base classes need to access this object as a const ref. I got away with a unique pointer as only concurrency needs to determine its life-spawn. But maybe a shared pointer is a more robust solution.  I could also consider virtual inheritance, but I need either some control on the initialization order, or to setup the class MPI_grouping outside of its constructor.

Reply:
There is an actul reason to use `unique_ptr`: concurrency needs to destroy the grouping (it frees an MPI communicator) before calling MPI_Finalize in the destructor. If the child owns a shared copy of the grouping this can't happen.